keps/sig-node/4383-dra-semantic-model-mode-switching/README.md

```diff
@@ -0,0 +1,898 @@
+      - label: name
+        op: GE
+        value: a
+      - label: memory
+        op: LT
+        value: 20Gi
+        type: Quantity
+      - label: driver
+        op: GT
+        value: v1.0.0
+        type: Semver
+```
+
+The underlying Go type is:
+```
+type Selector struct {
+    Label string
+    Op    SelectorOp // LT, LEQ, EQ, NEQ, GEQ, GT, IsSet, IsNotSet
+    Type  ValueType  // String (default), Quantity (= resource.Quantity), Semver (semantic version number)
+    Value string
+    And   []Selector
+    Or    []Selector
+}
+```
+
+
+### Capacity type
+
+The NodeResourceCapacity type already can list multiple resource
+instances. What this semantic model adds is a description of the different
+modes that each instance can be in.
+
+```
+// Capacity defines all modes available in one resource instance.
+type Capacity struct {
+    // Labels defined here become part of the set that a selector is matched against.
+    // Name and UID are just for information. The instance is identified inside
+    // the resource.k8s.io/NodeResourceCapacity via the ID that is set for each
+    // instance there.
+    ObjectMeta
+
+    Modes []Mode
+}
+
+// Mode is one of several modes that a instance can be in.
+type Mode struct {
+    // Labels defined here become part of the set that a selector is matched against,
+    // overwriting instance labels with the same name if there are duplicates.
+    // The name is for information. The UID must be unique among all modes of
+    // the instance.
+    ObjectMeta
+
+    // Items are numeric counters. They are independent of each other. When a claim
+    // gets allocated, what is left can be used for other claims.
+    Items []Item
+}
+
+type Item struct {
+    // Name identifies the kind of item that is being counted. Different modes
+    // may or may not have the same kind of items.
+    Name string
+
+    // Count is the total number of items in a certain mode.
+    Count int64
+}
+```
+
+### Claim parameter type
+
+```
+// Parameters defines what is needed for a claim.
+type Parameters struct {
+    Selector labels.Selector
+    Required []Item
+}
+```
+
+### Allocation result
+
+The following object gets serialized as JSON and stored in a claim's
+"status.allocation.resourcehandle.data" field:
+
+```
+// AllocationResult captures class and claim parameters and records which node
+// capacity instance was used. The JSON dump of this gets stored in the driver
+// handle field of an allocated claim.
+type AllocationResult struct {
+	// ClassParameters are the parameters from the time that the claim
+	// was allocated. They can be arbitrary setup parameters that are
+	// ignore by the counter controller.
+	ClassParameters runtime.RawExtension
+
+	// ClaimParameters are the parameters from the time that the claim was
+	// allocated. Some of the fields were used by the counter controller to
+	// allocated resources, others can be arbitrary setup parameters.
+	ClaimParameters runtime.RawExtension
```

Comment:
There's a discussion over in https://github.com/kubernetes/enhancements/pull/4384 about using `RawExtension`. Here it is unavoidable because the type of the claim parameters is the unknown vendor CRD. Same for the class parameters.